src/ninja.cc

```diff
@@ -1087,6 +1096,124 @@ int NinjaMain::ToolRestat(const Options* options, int argc, char* argv[]) {
   return EXIT_SUCCESS;
 }
 
+struct Compdb_Targets {
+  enum class Action {
+    display_help_and_exit,
+    emit_commands
+  };
+
+  Action action;
+  EvaluateCommandMode eval_mode = ECM_NORMAL;
+
+  std::vector<StringPiece> targets;
+
+  static Compdb_Targets parse(int argc, char* argv[]) {
+    //
+    // grammar:
+    //     ninja -t compdb-targets [-hx] target [targets]
+    //
+    Compdb_Targets ret;
+
+    // getopt_long() expects argv[0] to contain the name of
+    // the tool, i.e. "compdb".
+    argc++;
+    argv--;
+
+    // Phase 1: parse options:
+    {
+      optind = 1;  // see `man 3 getopt` for documentation on optind
+      int opt;
+      while ((opt = getopt(argc, argv, "hx")) != -1) {
+        switch (opt) {
+        case 'x':
+          ret.eval_mode = ECM_EXPAND_RSPFILE;
+          break;
+        case 'h':
+        default:
+          ret.action = Compdb_Targets::Action::display_help_and_exit;
+          goto fin;
+        }
+      }
+    }
+
+    // Phase 2: parse operands:
+    {
+      int targets_begin = optind;
+      int const targets_end = argc;
+
+      if (targets_begin == targets_end) {
+        Fatal("compdb-targets expects the name of at least one target");
+      } else {  // parse rules & targets (if any)
+        ret.action = Compdb_Targets::Action::emit_commands;
+        for (int i = targets_begin; i < targets_end; ++i)
+          ret.targets.push_back(argv[i]);
+      }
+    }
+
+  fin:
+    return ret;
+  }
+};
+
+
+void printCompdb(std::string const &directory, std::vector<Edge *> const &edges,
+                 const EvaluateCommandMode eval_mode) {
+  putchar('[');
+
+  bool first = true;
+  for (Edge* edge : edges) {
+    if (edge->is_phony() || edge->inputs_.empty())
+      continue;
+    if (!first)
+      putchar(',');
+    printOneCompdbObject(directory, edge, eval_mode);
+    first = false;
+  }
+
+  puts("\n]");
+}
+
+int NinjaMain::ToolCompilationDatabaseForTargets(const Options* options,
+                                                 int argc, char* argv[]) {
+  auto compdb = Compdb_Targets::parse(argc, argv);
+
+  switch (compdb.action) {
+  case Compdb_Targets::Action::display_help_and_exit: {
+    printf(
+        "usage: ninja -t compdb [-hx] target [targets]\n"
+        "\n"
+        "options:\n"
+        "  -h     display this help messsage\n"
+        "  -x     expand @rspfile style response file invocations\n");
+    return 1;
+  }
+
+  case Compdb_Targets::Action::emit_commands: {
+    InEdge_Collector inEdge_collector;
+
+    for (auto target_arg : compdb.targets) {
+      std::string err;
+      Node* node = CollectTarget(target_arg.begin(), &err);
```

Comment:
There is no guarantee that `StringPiece::begin()` is zero-terminated, so I think there may be a bug here.
Use `target_arg.AsString().c_str()` instead to fix it.

Reply:
thank you for catching this!

Reply:
wait; on second thought: in this case, `target_arg.begin()` (a.k.a. `target_arg.str_`) _is_ guaranteed to be null-terminated, because it's the same pointer value as one of the elements of `argv` from `main()`; e.g. if there are no options, then the first `target_arg.str_` ==  `argv[3]` (and if that's not null-terminated, then something weirdly bad must have happened much earlier).

also, every correct use of `StringPiece` depends on knowing the provenance of the string pointed-to. Since we need to know the provenance anyway, and in this case knowing the provenance also comes with the knowledge that the string is null-terminated, it seems ok to depend on that...?

Reply:
In _general_, StringPiece::begin() is not zero-terminated. In this specific line, it happens to be the case though, because of the way compdb.targets is _currently_ built, but this is not visible from the current code's context.

It is easy to imagine that someone else later changes how combd.targets is populated (for example by expanding the list from a response file, or to support the `target^` syntax), and then you will have an unexpected bug.

Maybe StringPiece is not the right type for the targets vector, and that a `const char*` type would be better because it exposes the fact that the strings are zero-terminated? Or just use a vector of strings?